### PR TITLE
Introduce GcscliBlobstoreClient to bosh-director

### PR DESCRIPTION
**Pull request by Everlag**
GcscliBlobstoreClient is the S3cliBlobstoreClient equivalent
implementing support for natively using GCS as a bosh-director
blobstore.

This includes basic unit and functional tests. Unit tests are
at par with those running against S3cliBlobstoreClient while
the functional tests are a subset. At least some of the general_s3
suite are blocked on bosh-gcscli supporting a read-only mode,
requested by setting credentials_source to 'none'.

bosh-gcscli binary has not yet been included in packages as a
dependency as it is still open to changing significantly across
working commits.

### CODE REVIEW

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (johnsonj) [inline src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb:25]: nit: extra new line

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (johnsonj) [inline src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: `bosh-gcscli binary`, go just a detail

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (johnsonj) [inline src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: you can drop support for Ruby 1.8, looks like they use 2.3.1

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (johnsonj) [inline src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: magic number?

**Review by johnsonj — COMMENTED** [review]
just skimmed a few quick comments, i'll leave this to mkjelland@

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (Everlag) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: Convention established in s3cli, [implementation here](https://github.com/cloudfoundry/bosh-gcscli/blob/master/main.go#L169)

Comment (johnsonj) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: but can you use a constant here?

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (Everlag) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb:25]: Done

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (Everlag) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: Done

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (Everlag) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: Introduced constant in bbd3b0b

File: jobs/director/templates/director.yml.erb.erb
Comment (mkjelland) [inline jobs/director/templates/director.yml.erb.erb:192]: Do the bucket_name/credentials_source/storage_class params need to get passed in from this file? It looks like they get passed in if the provider is s3 on lines 133-139.

File: jobs/director/templates/director.yml.erb.erb
Comment (mkjelland) [inline jobs/director/templates/director.yml.erb.erb:195]: How does the bosh-gcscli binary get pulled in? Will it be added to the packages folder?

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (mkjelland) [inline src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: What is the difference between this if statement and having a second param on the @options.fetch call on line 32?

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (mkjelland) [inline src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: I like how you use EXIT_CODE_OBJECT_NOTFOUND instead of hard coding the 3 here. I'd recommend using the same pattern here for the 0.

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (mkjelland) [inline src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb:33]: What happens if credentials_source is none until the bosh-gcscli supports a read-only mode?

Comment (Everlag) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb:33]: bosh-gcscli now supports read-only mode.

File: jobs/director/templates/director.yml.erb.erb
Comment (Everlag) [reply jobs/director/templates/director.yml.erb.erb:195]: Yes, the packages folder. That's one of the last things to do as it involves Pivotal people pushing to a blobstore.

File: jobs/director/templates/director.yml.erb.erb
Comment (Everlag) [reply jobs/director/templates/director.yml.erb.erb:192]: Looks like I missed this. Is it fine to defer making this change until bosh-gcscli stabilizes the configuration file a bit more?

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (Everlag) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: Good catch, removed this.

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (Everlag) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb]: Done.

File: jobs/director/templates/director.yml.erb.erb
Comment (mkjelland) [reply jobs/director/templates/director.yml.erb.erb:195]: Ok, that makes sense. Is this blocked on merging until the it is there?

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (mkjelland) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb:33]: [s3_spec.rb](https://github.com/Everlag/bosh/blob/e4d1d19ed80e8c50ff7bf39ede8165b32fc166bd/src/bosh-director/spec/functional/s3_spec.rb) has some tests for the read only case. Do you think it makes sense to add those in gcs_spec.rb now that this is supported?

File: jobs/director/templates/director.yml.erb.erb
Comment (mkjelland) [reply jobs/director/templates/director.yml.erb.erb:192]: What will happen if those params are not passed in? Will it default to read-only mode?

Comment (Everlag) [reply jobs/director/templates/director.yml.erb.erb:192]: If I had to guess, I'd say it would error out due to an empty bucket name. Running an adapted BATs, which I've yet to, would probably confirm this.

File: jobs/director/templates/director.yml.erb.erb
Comment (Everlag) [reply jobs/director/templates/director.yml.erb.erb:195]: Yes, I believe a note in our PR following the style of this [PR](https://github.com/cloudfoundry/bosh/pull/1494) is the way to go.

(By being blocked on merging, I assume you mean into the main cloudfoundry/bosh repo rather than this PR)

File: src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb
Comment (Everlag) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb:33]: Read-only tests are included in an additional, smaller PR into this fork. Given that its a 61 line addition, I can include them here if you think that's a reasonable amount to add.

Comment (mkjelland) [reply src/bosh-director/lib/bosh/blobstore_client/gcscli_blobstore_client.rb:33]: Ok, a separate PR seems fine.

File: jobs/director/templates/director.yml.erb.erb
Comment (mkjelland) [reply jobs/director/templates/director.yml.erb.erb:192]: This is also something that will need to happen before merging into the main bosh repo, but if you want to wait until you know what the params will be that is alright. The unit tests will need to be updated once you know what the params will be too.

**Review by mkjelland — APPROVED** [review]

File: jobs/director/templates/director.yml.erb.erb
Comment (Everlag) [reply jobs/director/templates/director.yml.erb.erb:192]: I've added a note to my personal docs to include this before trying to merge.

**Pull request merged**
